site/content/en/docs/build-clusters.md

```diff
@@ -0,0 +1,181 @@
+---
+title: "GKE Build Clusters"
+description: >
+  Sometimes you might want to run all of your jobs in a GKE cluster that is separate from Prow (perhaps running in your own GCP project). What you need is a dedicated "build cluster".
+weight: 3
+---
+
+*Note: This page discusses build clusters that use GKE. Technically speaking, a
+build cluster could be any Kubernetes cluster (not just GKE) because the only
+thing Prow needs is the ability to authenticate as a Kubernetes Service Account
+with cluster-admin role permissions to the build cluster.*
+
+## Overview
+
+By default Prow will schedule jobs in the same Kubernetes cluster as itself, but
+under the `test-pods` namespace. Setting up a separate build cluster will allow
+you to schedule jobs into a different Kubernetes cluster altogether. When
+everything's set up, Prow's component will schedule jobs into your build cluster
+(as shown in the [Arhictecture diagram](../overview/architecture)) instead of
+its own. This way, you can "bring your own build cluster" to Prow to make it
+scale however you see fit.
+
+For convenience we use the terms KSA and GSA, where **KSA** means **Kubernetes
+Service Account** and **GSA** means **GCP (IAM) Service Account**.
+
+## Running the build cluster setup script
+
+The Prow source repo comes with a default
+[create-build-cluster.sh](https://github.com/kubernetes/test-infra/blob/master/prow/create-build-cluster.sh)
+script which allows you to create a new GKE cluster with the intent of giving
+the [Kubernetes Prow instance](https://prow.k8s.io) access to it. Because there
+are different Prow instances and each instance has its own default settings
+(esp. for permissions), each instance has its own such script, forked from the
+default one. For example, [Google's OSS Prow instance](https://oss.gprow.dev)
+has its own script
+[here](https://github.com/GoogleCloudPlatform/oss-test-infra/blob/master/prow/oss/create-build-cluster.sh).
+
+The scripts all have prompts and ask you various questions to set everything up.
+If everything proceeds smoothly, then there's nothing more for you to do (you're
+ready to start writing Prow jobs that use your cluster). Below is a discussion
+of the overall process to demystify what goes on behind the scenes.
+
+## How does Prow actually use your build cluster?
+
+There are two requirements:
+
+1. Prow must be able to schedule jobs into your build cluster.
+2. The jobs themselves must be able to upload artifacts to the GCS bucket used
+by [Deck](../components/deck), in order to report job status (e.g., "passing" or
+"failing").
+
+We look at both requirements below.
+
+### Let Prow schedule your jobs into your build cluster
+
+Prow is a Kubernetes cluster. So is your build cluster. In order for Prow to
+schedule jobs (i.e., create Kubernetes pods) into your build cluster, it must be
+able to authenticate as a KSA defined in your build cluster that has a
```

Comment:
KSA is not a hard requirement, it's just an easier way to generate kubeconfig for authenticating with a cluster. Being able to manage `Pod` resource on the build cluster might be more accurate, for more detail see gencred code

Reply:
But what about https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L78-L79 ? Seems to contradict what you just said. Unless there's other `gencred` codepaths involved here.

I will simply include your words verbatim into the doc.

Reply:
I'm confused because this line seems to condtradict what you're saying: https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L78

But also, I'll include your comment in the doc directly, because it will be faster than me reading through gencred.

Reply:
I don't think there is a conflict here. The point I'd like to make, is that "it must be able to authenticate as a KSA" is not an accurate statement, what Prow really must have is "it must be able to `kubectl get pods`, `kubectl create pods`, `kubectl * pods` etc", and other than creating a KSA, there are multiple other ways this can be achieved. At least gencred supports two different ways https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/cmd/gencred/main.go#L236

Reply:
> I'm confused because this line seems to condtradict what you're saying: https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L78
> 
> But also, I'll include your comment in the doc directly, because it will be faster than me reading through gencred.

Well please hold off a few minutes if this is what you plan to do, I'll need to read through gencred code to make sure that I didn't miss something especially the RBAC stuff

Reply:
> Well please hold off a few minutes if this is what you plan to do, I'll need to read through gencred code to make sure that I didn't miss something especially the RBAC stuff

So according to https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L36 that this KSA is a "cluster-admin" super user. I'm not entirely convinced that it's a requirement, feels like a security hole to me. @cjwagner , do you know what resource Prow needs to manage other than `Pod` resource on build clusters?